packages/utreexo/src/stump/accumulator.cairo

```diff
@@ -1,13 +1,48 @@
 use crate::stump::state::UtreexoStumpState;
 use crate::stump::proof::{UtreexoBatchProof, UtreexoBatchProofTrait};
+use crate::parent_hash;
+
+use utils::bit_shifts::{shr_u64};
 
 #[generate_trait]
 pub impl StumpUtreexoAccumulatorImpl of StumpUtreexoAccumulator {
     /// Adds multiple items to the accumulator.
     fn add(self: @UtreexoStumpState, hashes: Span<felt252>) -> UtreexoStumpState {
-        // TODO: check if vanilla implementation is compatible with rustreexo, add tests
-        // https://github.com/mit-dci/rustreexo/blob/6a8fe53fa545df8f1a30733fa6ca9f7b0077d051/src/accumulator/stump.rs#L252
-        *self
+        let mut leaves = *self.num_leaves;
+
+        // This array will hold state of roots after adding new hashes
+        // IMPORTANT: This line copies all roots from the current state
+        // to the new array
+        let mut new_roots: Array<Option<felt252>> = (*self.roots).into();
+
+        for add in hashes {
+            // We will use new_roots_span to iterate over the roots that we
+            // will merge (thus remove from the state) with the new hash.
+            let mut new_roots_span = new_roots.span();
+            let mut h = 0;
+            let mut to_add = *add;
+
+            // This is similar to the vanilla algorithm
+            while shr_u64(leaves, h) % 2 == 1 {
```

Comment:
Smth like this should work:
```
let (mut next_row_len, mut root_expected) = DivRem(leaves, 2);
while root_expected == 1 {
    let (q, r) = DivRem(next_row_len, 2);
    next_row_len = q;
    root_expected = r;
}
```

Reply:
Thank you save 30% of gas according tests.